cpp/src/link_analysis/hits_impl.cuh

```diff
@@ -0,0 +1,201 @@
+#include <cugraph/prims/copy_v_transform_reduce_in_out_nbr.cuh>
+#include <cugraph/prims/count_if_v.cuh>
+#include <cugraph/prims/reduce_v.cuh>
+#include <cugraph/prims/row_col_properties.cuh>
+#include <cugraph/prims/transform_reduce_v.cuh>
+
+// TODO : delete
+#include <string>
+
+namespace cugraph {
+namespace detail {
+template <typename GraphViewType, typename result_t>
+void normalize(raft::handle_t const& handle,
+               GraphViewType const& graph_view,
+               result_t* hubs,
+               raft::comms::op_t op)
+{
+  auto hubs_norm = reduce_v(handle,
+                            graph_view,
+                            hubs,
+                            hubs + graph_view.get_number_of_local_vertices(),
+                            identity<result_t>(op),
+                            op);
+  thrust::transform(handle.get_thrust_policy(),
+                    hubs,
+                    hubs + graph_view.get_number_of_local_vertices(),
+                    thrust::make_constant_iterator(hubs_norm),
+                    hubs,
+                    thrust::divides<result_t>());
+}
+
+template <typename GraphViewType, typename result_t>
+std::tuple<result_t, size_t> hits(raft::handle_t const& handle,
+                                  GraphViewType const& graph_view,
+                                  result_t* const hubs,
+                                  result_t* const authorities,
+                                  result_t epsilon,
+                                  size_t max_iterations,
+                                  bool has_initial_hubs_guess,
+                                  bool normalize,
+                                  bool do_expensive_check)
+{
+  using vertex_t = typename GraphViewType::vertex_type;
+  static_assert(std::is_integral<vertex_t>::value,
+                "GraphViewType::vertex_type should be integral.");
+  static_assert(std::is_floating_point<result_t>::value,
+                "result_t should be a floating-point type.");
+  static_assert(GraphViewType::is_adj_matrix_transposed,
+                "GraphViewType should support the pull model.");
+
+  auto const num_vertices = graph_view.get_number_of_vertices();
+  result_t hubs_error{std::numeric_limits<result_t>::max()};
+  size_t final_iteration_count{max_iterations};
+
+  if (num_vertices == 0) { return std::make_tuple(hubs_error, final_iteration_count); }
+
+  CUGRAPH_EXPECTS(epsilon >= 0.0, "Invalid input argument: epsilon should be non-negative.");
+
+  // Check validity of initial guess if supplied
+  if (has_initial_hubs_guess && do_expensive_check) {
+    auto num_negative_values =
+      count_if_v(handle, graph_view, hubs, [] __device__(auto val) { return val < 0.0; });
+    CUGRAPH_EXPECTS(num_negative_values == 0,
+                    "Invalid input argument: initial guess values should be non-negative.");
+    detail::normalize(handle, graph_view, hubs, raft::comms::op_t::SUM);
```

Comment:
Should we normalize even when the input argument ```normalize``` is false? And ```do_expensive_check = true``` shouldn't change the program behavior. If the input is expected to be normalized, users should provide normalized values. If not, we should normalize outside the ```if (do_expensive_check)```.

Reply:
How about this.,
We check if the input is normalized inside this block and throw logical error if it is not.

We normalize the input outside of this check block when has_initial_hubs_guess and normalize is true.